src/Spectre.libDataset/Dataset.h

```diff
@@ -0,0 +1,192 @@
+/*
+* Dataset.h
+* Data-owning implementation of Dataset interfaces.
+*
+Copyright 2017 Maciej Gamrat, Grzegorz Mrukwa
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include <vector>
+#include "IDataset.h"
+#include "Exception.h"
+
+namespace Spectre::libDataset
+{
+    /// <summary>
+    /// Data-owning implementation of dataset interfaces.
+    /// </summary>
+    template<typename DataType, typename SampleMetadata, typename DatasetMetadata>
+    class Dataset : public IDataset<DataType, SampleMetadata, DatasetMetadata>
```

Comment:
While I agree on *usual* methods, I think we should leave the interface consistent with *C++-like* whenever mimicking it, so this would apply e.g. to: `size`, `empty`.